File: java/MaximumSubmatrix.java
public class Solution {
    // https://www.lintcode.com/problem/maximum-submatrix/description
    private int maxSum;
    private int[] rowPrefixSum;
    private int m;
    private int n;
    
    private void computePrefixSum(int colIndex, int[][] matrix) {
        for(int i=0; i < m; ++i) {
            rowPrefixSum[i] += matrix[i][colIndex];
        }
    }
    
    private void maxSubarray() {
        int currentSum = 0;
        for(int i = 0; i < m; ++i) {
            currentSum = Math.max(currentSum + rowPrefixSum[i], rowPrefixSum[i]);
            maxSum = Math.max(currentSum, maxSum);
        }
    }
    
    /**
     * @param matrix: the given matrix
     * @return: the largest possible sum
     */
    public int maxSubmatrix(int[][] matrix) {
        // write your code here
        m = matrix.length;
        if(m == 0) return 0;
        n = matrix[0].length;
        if(n == 0) return 0;
        
        rowPrefixSum = new int[m];
        
        for(int leftColIndex = 0; leftColIndex < n; ++leftColIndex) {
            for(int rightColIndex = leftColIndex; rightColIndex < n; ++rightColIndex) {
                computePrefixSum(rightColIndex, matrix);
                maxSubarray();
            }
            rowPrefixSum = new int[m];
        }
        
        return maxSum;
    }
}

File: python/maximum-submatrix.py
class Solution:
    """
    @param matrix: the given matrix
    @return: the largest possible sum
    """
    def maxSubmatrix(self, matrix):
        # write your code here
        # turn it into a small problem
        if not matrix or not matrix[0]:
            return 0
        
        n, m = len(matrix), len(matrix[0])
        maxi = -sys.maxsize + 1
        
        for lowerbound in range(n):
            # because it only depend on the last sum
            lastSums = [0 for _ in range(m)]
            for i in range(lowerbound, n):
                for j in range(m):
                    lastSums[j] += matrix[i][j]
                maxi = max(maxi, self.maxSubArray(lastSums))
        
        return maxi
        
    def maxSubArray(self, array):
        rollingSum = mini = 0
        maxi = -sys.maxsize + 1
        
        for num in array:
            rollingSum += num
            maxi = max(maxi, rollingSum - mini)
            mini = min(mini, rollingSum)
        
        return maxi
    